### ISSUE TITLE
Plugin Errors: Allow option to change resolver return type 

### ISSUE THREAD

**Issue by jasonkuhrt** [hayes/pothos#1102]
We do not use `throw` in our codebase. We only return errors. I would like an option on the errors plugin https://pothos-graphql.dev/docs/plugins/errors to use a "return mode" wherein the resolver return types are made into unions wherein the errors (backing type) can be returned.

Using `throw` is inferior because we can `throw` _anything_ without _any_ type safety.

If we return an error that is not expected by a resolver, we want to know about it.

**Comment by hayes** [comment]
Sorry, this issue slipped through my notifications, just seeing it now.

This definitely makes sense, but I'm snuggling to figure out how this can be implemented.   The issue is that there isn't a good way for a plugin to change the expected return type of a resolver by through the options it adds to field methods.

It would definitely be possible to add something like `t.errorField` that allows you to return the errors defined for the field, but this wouldn't be composable with other plugins (eg. you wouldn't be able to return errors in t.connection, unless we also added a new `t.errorConnection` field).  

the core issue is that all methods like t.field have a set number of generic "slots" that are used to infer specific things, and can be used to derive data for additional options, but being able to infer more data from options generally isn't possible without defining custom methods that have additional options.

The closest thing that comes to mind in other plugins is the `t.withAuth` method from the scope-auth plugin which changes the context type in resolvers based on the auth (eg `t.withAuth({ loggedIn: true }).field({ ..., resolve: (_, args, ctx) => ctx.user })` where the logged in scope is defined as having a context object where the user is always set.   I'm not sure if this would be a good pattern, but it might be possible to implement something like `t.withErros([AuthError]).field({ ..., resolve: () => { if (notLoggedIn()) { return new AuthError(...) } ... })`

Definitely open to suggestions on how to make this work.  I agree that it would make sense to allow this, but I am unsure about the best way to implement it given the constraints in the type system

**Comment by jasonkuhrt** [comment]
In Nexus we "cheated" with the codegen approach. I am doing some deep-ish TS stuff with my more recent project at https://github.com/jasonkuhrt/molt and so have been sharpening my skills, most recently leveraging HKTs for some features. All to say that I would be curious to take a look at how it could be achieved in the Pothos types. Time is my limiting factor but I've taken a note about maybe looking into this if I ever have time. Thanks for the great tool @hayes!

**Comment by hayes** [comment]
Nice, would love to get your take on this!  More than happy to help out however I can, or answer any questions you have about how types work.

Something I am exploring for 4.0 is having the base field options entirely replaceable.  This is intended for allowing easy backwards compatibility with v3, so you just switch an option, and you get the field options as they were for v3, but I am also thinking about grafast which uses plans instead of resolvers.   This potentially could also work for the errors plugin.  This still wouldn't fully composable (You can opt into using the field options from the error plugin, so that resolvers can return any defined error types, but you would be opting into that over using the options from v3 or the Grafast plan options.  You could potentially add in a `v3WithErrors` option or something, but you don't get composition out of the box.

example from a conversation about grafast:

```ts
interface FieldOptionsByVersion<Type, Args, ...>  {
  v3: { resolve: Resolver<...>, nullable?: Nullable }
  v4: { resolve: Resolver<...>, nonNullable?: Nullable }
  // Interface extended via plguin but inlined here
  Grafast: { nonNullable?: Nullable } & ({ plan: Plan<...> } | { resolve: Resolver<...> } )
}

field<Type, Args, ...>(options: FieldOptionsByVersion<Type, Args, ...>) { ... }
```


I am also not entirely opposed to making returnable errors part of Pothos core.  You'd still need to use the errors plugin to have GraphQL types representing your errors, but we could do something like:

```ts
const builder = new Builder<{
  // Errors that are allowed to be returned by default (not specific to errors plugin) defaults to `never`
  Errors: AuthError | RateLimitError
}>({...})

// psudo code, oversimplified/wrong in a bunch of ways
interface BaseFieldOptions<Type, Args, OtherCurrentGenerics, Errors = Types['Errors']> {
  resolve: (parent: Type, args: Args, ctx: Types['Context'], info: GraphQLResolveInfo) => MaybePromise<ReturnType | Errors>
}

// extend in errors plugin
interface BaseFieldOptions<Type, Args, OtherCurrentGenerics, Errors = Types['Errors']> {
  errors: { type:s Errors[] }
}
```
Basically this adds a new Generic specifically for returned errors.  You can globally allow specific errors to be returned from any field, then if your field defines errors, the return type would be allowed to return those specific errors.

**Comment by jasonkuhrt** [comment]
A little off topic: Cool didn't know about Grafast. On my team at Prisma I'm currently prototyping a data-loader-as-Prisma-extension which will, within a request lifecycle, dedupe all records fetched more than once. With rich relationship knowledge provided for free by metadata generated from the Prisma Client generator, the amount of automated performance increase we can provide for zero learning curve looks promising. Of course, it does not help on the initial fetch of records. Anyways, it's definitely more specific than Grafast but I also with a much lower learning curve (none, basically). I see you (or community) have made some cool Prisma plugins for Pothos. I'm excited to give them a try as we advance on our GraphQL API work. If our prisma extension shows worth and we open source it, maybe it would be worth mentioning it on the Prisma parts of the Pothos docs (e.g. a "tip" callout).

**Comment by hayes** [comment]
Definitely excited to see what you come up with there!  I think there's a good chance it will be easy to integrate with what I am doing with the prisma plugin already, and could potentially be integrated directly with the plugin to make things even easier.

The current prisma plugin already gives pretty good results with very low user effort.  Basically it does the mapping of query to select/includes for the top level resolver, but is a little more careful about it than other solutions, so you won't run into any issues if you have something like `type Author { drafts: [Post], posts: [Post] }` where both `drafts` and `posts` are using the same prisma relation with different queries.

from a user perspective this might be something like:
```ts
builder.queryField('author', t => t.prismaField({
  type: 'Author',
  args: { id: t.arg.id({ required: true }) },
  resolve: (query, _, args) => prisma.author.findUnique({ where: { id: args.id }, ...query })
}))

builder.prismaObject('Author', {
  fields: t => ({
    posts: t.relation('posts', {
      query: { where: { draft: false } }
    }),
    drafts: t.relation('posts', {
      query: {
        orderBy: { modifiedAt: 'desc' },
        where: { draft: true }
      }
    }),
  })
})
```

When only 1 of drafts/posts are selected in that GQL query, that gets included into query that gets passed into the `queryField`. When they are both selected, only first get's included, and when the resolver runs for the other field, it basically uses an internal dataloader to re-load the the Author object with the remaining posts (basically `prisma.author.findUnique({ select: { posts: { where: { draft: true } })`.   There are a few tricks in the to ensure that Prisma will properly batch these findUnique calls, and the dataloader ensures that we add selects for as many fields as possible, rather than just having this done for each field that couldn't be part of the initial query.   We also do the same mapping to select any sub-selections that can be loaded as part of these data-loaded calls.  It gets a little complex, but I haven't found many places where a manual optimization would have been better.

The plugin also supports defining fields with arbitrary additional selects/includes that it depends on so you can do something like:


builder.prismaObject('Author', {
  hasMorePostsThanComments: t.field({
    select: {
      _count: { posts: true, comments: true }
    },
    resolve:  author => author._count.posts > author._count.comments,
  })
})
```

and the selections will be added into the parent query, and the types properly infer the added selections on the parent object. 

I think if the dataloader you are talking about can support data-loading in a way that works with these relation based queries that would be really cool.   The way you descrived it, it might just work without any additional integration on our end required

**Comment by jasonkuhrt** [comment]
Thanks for sharing @hayes! What I was expecting was, at a high level:

> Convert incoming GraphQL document into the most compact/de-duped Prisma Client query representation.

The general idea of:

1. Look at incoming document
2. Analyze the relationship between what GraphQL objects are selected and the backing prisma model(s)
3. Write the most efficient/natural `prisma.*.findUnique/findMany/...({ include: { ... }})` query based on analysis output of (2) (basically something as good as what could be written by developer)

The above seems invaluable to me for the following reasons:

1. Fewer round trips to the database for relations/dependencies.
2. Optimize mapping of: (1) GraphQL document to (2) Prisma operation, in turn allowing Prisma to optimize the mapping of (1) prisma operation to (2) database query (SQL, mongodb query, etc. etc.)

(2) Above is exciting because for no effort from Pothos user's can see their performance get better and better as Prisma itself improves its query builder system. For example the latest release of Prisma will now be making real SQL JOINs!

You mention something about if a single Prisma query's includes repeat the fetching of some records. I believe deduping/optimizing this should be Prisma Client's internal responsibility, built into its core. As of today, I suspect it's not yet optimized to do anything smart there. However, maybe a reason for that is because anyways the database (e.g. Postgres) is already great at effectively making those dedupes (if only Prisma would leverage native JOINs, which as I just mentioned, they are now beginning to). That said, maybe there is something to be said about reducing what is sent over the wire, too. Meaning: having Prisma Client do the "injection" of duplicated record fetches in memory on application process side.

IIUC though you're doing something a bit different which is introducing some cache to dedupe `findUnique` etc. calls?

If so, I think you're doing some work there that what we're building would automatically take care of as well. But nothing that is incompatible of course!

The thing that we're building does not, and never will, take care of the aforementioned optimal mapping between GraphQL document and backend Prisma client query.

Thinking about this, the greatest potential I see for compositional/decoupled optimal performance building blocks re GraphQL+Prisma is:

1. GraphQL Document <> Prisma Operation
  Component: Pothos Prisma Plugin
  Goal: cut away as much fat as possible from the outset
2. Prisma Extension: Prisma Data Loader (dedupe same-record fetches)
  Component: Prisma Extension
  Goal: pick up slack that (1) couldn't handle for whatever reason 
3. Prisma Operation <> Database
  Component: Prisma (core)
  Goal: write the most efficient possible SQL/Mongo query/... etc.

Regardless of the breadth of optimizations you're making/aiming to make with the Prisma Pothos plugin, I don't think anything it does do/could do would "conflict" with our data loader extension. At "worst" it would just be duplicating some things it also does, but with no visible effect, given cache behaviours behaving same, idempotent, etc.

I guess what appeals to me about the above optimization separations is that each layer can focus on doing what it does as best as it can, with zero coordination. And indeed each layer there brings a critical element of performance, unique in an upward facing way (each layer can theoretically make all the optimizations of lower layers but assuming strict scope separation nothing "above" itself given it does not know anything about its usage context.

wdyt?

**Comment by hayes** [comment]
I think we are very aligned on the separation of concerns/scope of optimizations!  What you are describing is exactly how the Pothos plugin operates.

I was a little in-accurate when I said we use a dataloader in the plugin.  Really what we have is a batching/scheduling mechanism that does 2 things:

1. Enables us to add as much into single `prisma.*.findUnique` call
2. Schedules batches of `prisma.*.findUnique` calls for a type into the same "tick" based on type so that prismas n+1 optimization kicks in properly (so they get batched into one `select ... where id in (...)`). 

The scheduling piece turns out to be really important, because prismas internal batching bails out if you make prisma queries for multiple models in the same tick, so we ensure that the prisma calls that get grouped by type onto different ticks

**Comment by hayes** [comment]
The 4.0 release has a new way to extend how resolve works in the builder that could be used to implement something like this.  If this is still something you are interested in, I could probably put together an example of adding errors to accepted Resolver return types (there might be some exceptions where this doesn't work as expected)

**Comment by jasonkuhrt** [comment]
@hayes Yeah this would be a significant DX improvement for us still!

**Comment by hayes** [comment]
I have not tested this extensively, but I think this works pretty well
```ts
import SchemaBuilder, {
  FieldNullability,
  InputFieldMap,
  InputShapeFromFields,
  ListResolveValue,
  MaybePromise,
  Resolver,
  SchemaTypes,
  ShapeFromTypeParam,
  TypeParam,
} from '@pothos/core';

type ResolveShapeWithErrors<Shape, Errors> = [Shape] extends [
  readonly (infer Item)[] | null | undefined,
]
  ? ListResolveValue<Shape, Errors | Item, unknown>
  : MaybePromise<Errors | Shape>;

// Should match what you pass as default errors when setting up the builder
type DefaultErrors = CustomError1;

declare global {
  export namespace PothosSchemaTypes {
    export interface InferredFieldOptions<
      Types extends SchemaTypes,
      ResolveShape = unknown,
      Type extends TypeParam<Types> = TypeParam<Types>,
      Nullable extends FieldNullability<Type> = FieldNullability<Type>,
      Args extends InputFieldMap = InputFieldMap,
      ResolveReturnShape = unknown,
    > {
      ResolveWithErrors: {
        errors?: {
          types?: (new (...args: any[]) => ResolveReturnShape)[];
        };
        /**
         * Resolver function for this field
         * @param parent - The parent object for the current type
         * @param {object} args - args object based on the args defined for this field
         * @param {object} context - the context object for the current query, based on `Context` type provided to the SchemaBuilder
         * @param {GraphQLResolveInfo} info - info about how this field was queried
         */
        resolve: Resolver<
          ResolveShape,
          InputShapeFromFields<Args>,
          Types['Context'],
          ResolveShapeWithErrors<
            ShapeFromTypeParam<Types, Type, Nullable>,
            | DefaultErrors
            | (unknown extends ResolveReturnShape
                ? never
                : ResolveReturnShape extends infer T
                  ? T
                  : never)
          >
        >;
      };
    }
  }
}

class CustomError1 extends Error {
  prop1 = 'test';
}

class CustomError2 extends Error {
  prop2 = 'test';
}

const builder = new SchemaBuilder<{
  InferredFieldOptionsKind: 'ResolveWithErrors';
}>({
  errors: {
    defaultTypes: [Error],
  },
});

builder.queryType({
  fields: (t) => ({
    hello: t.string({
      errors: {
        types: [CustomError2],
      },
      resolve: () => {
        if (Math.random() < 0.2) {
          return new CustomError1('Random error');
        }

        if (Math.random() < 0.3) {
          return new CustomError2('Random error');
        }

        return 'Hello, world!';
      },
    }),
  }),
});

builder.subscriptionType({
  fields: (t) => ({
    hello: t.string({
      // eslint-disable-next-line @typescript-eslint/require-await
      async *subscribe() {
        yield 'Hello';
        yield 'World';
      },
      resolve: (message) => {
        if (Math.random() > 0.5) {
          return new CustomError1('Random error');
        }

        return message;
      },
    }),
  }),
});

```

**Comment by hayes** [comment]
The biggest drawback with this is that you get slightly worse type errors when returning the wrong thing from a list resolver (the normal resolve uses `ResolveReturnShape` is used to improve the errors, and is repurposed here to infer the error types instead)

**Comment by hayes** [comment]
Let me know how this works for you once you have a chance to try it out, gonna close this issue for now

**Comment by jasonkuhrt** [comment]
@hayes amazing thank you. I've got a thread in our internal Slack about this going now :)

**Comment by hayes** [comment]
awesome, looking forward to hearing how it goes
